code/game/objects/items/devices/oxycandle.dm

```diff
@@ -0,0 +1,73 @@
+/obj/item/device/oxycandle
+	name = "oxygen candle"
+	desc = "A steel tube with the words 'OXYGEN - PULL CORD TO IGNITE' stamped on the side. A small label warns against using the device underwater"
+	icon = 'icons/obj/device.dmi'
+	icon_state = "oxycandle"
+	item_state = "oxycandle"
+	w_class = ITEMSIZE_SMALL // Should fit into internal's box or maybe pocket
+	var/target_pressure = ONE_ATMOSPHERE
+	var/datum/gas_mixture/air_contents = null
+	var/volume = 5600 // One tile has 2500 volume of air, so two tiles plus a bit extra
+	var/on = FALSE
+	var/activation_sound = 'sound/items/flare.ogg'
+	light_color = LIGHT_COLOR_FLARE
+	uv_intensity = 50
+	var/brightness_on = 2 // Moderate bright.
+	light_power = 2
+	action_button_name = null
+
+/obj/item/device/oxycandle/attack_self(mob/user)
+	if(!on)
+		user << "<span class='notice'>You pull the cord and witness chemical reaction turn into fire that smells very refreshing.</span>"
+		light_range = brightness_on
+		on = TRUE
+		update_icon()
+		playsound(src.loc, activation_sound, 75, 1)
+		air_contents = new /datum/gas_mixture()
+		air_contents.volume = 200 //liters
+		air_contents.temperature = T20C
+		var/list/air_mix = list("oxygen" = O2STANDARD * (target_pressure * air_contents.volume) / (R_IDEAL_GAS_EQUATION * air_contents.temperature),
+		 						"nitrogen" = N2STANDARD *  (target_pressure * air_contents.volume) / (R_IDEAL_GAS_EQUATION * air_contents.temperature))
+		air_contents.adjust_multi("oxygen", air_mix["oxygen"], "nitrogen", air_mix["nitrogen"])
+		START_PROCESSING(SSprocessing, src)
+
+// Process of Oxygen candles releasing air. Makes 200 volume of oxygen and nitrogen mix
+/obj/item/device/oxycandle/process()
+	if(!loc)
+		return
+	var/turf/pos = get_turf(src)
+	if(volume <= 0 || istype(pos, /turf/simulated/floor/beach/water) || istype(pos, /turf/unsimulated/beach/water))
```

Comment:
Hacky and ugly. Surely there's a better way.

Reply:
Please explain how is this hacky and ugly? These are same conditions that need it to burn out.

Reply:
This is easy - we have set volume of chemicals inside, they burn at rate of 200. Once we run out of chemicals we burnt whole candle. Or if we are in the water - sodium chlorinate would dissolve - hence burning it out as well

Reply:
There generally shouldn't be explicit checks for turf types except for special cases like space. Could check if the turf has reagents & if that reagent list contains water.

Reply:
@Lohikar water turfs in an of themselves are, well, hack-y. By virtue of us not having an actual fluid system. They don't actually have water in their reagents.

Reply:
@skull132 I'm aware that they're hacky, I just thought that PoZe's recent changes to water turfs had added water reagent persistently. 

Either way, point stands. Explicit type checks like this are ugly and should be used sparingly if at all.

Reply:
@Lohikar @skull132 That is a good point. My recent changed to water thought call transfer of reagents between turf to a mob or atom. Let me check my code and see if it actually makes turfs have water.

Reply:
You were right, my recent updates does add reagents to water turfs. By default turf do not have water as the reagent, but if any atom enters it - it calls proc that adds water to the turf. So I could check for it. Only problem would be that with other type of water which is unsimulated, it does not have a reagent

Reply:
There is a better way - Add a TURF_IS_WET flag.